ext/standard/array.c

```diff
@@ -1745,7 +1745,7 @@ static zend_long php_extract_ref_if_exists(zend_array *arr, zend_array *symbol_t
 				continue;
 			}
 			if (zend_string_equals_literal(var_name, "this")) {
-				zend_throw_error(NULL, "Cannot re-assign $this");
+				zend_value_error("Cannot re-assign $this");
```

Comment:
Not really a value error, and should be consistent with the corresponding engine error (which uses Error).